### PR TITLE
refactoring: raii-ify parser, make some err locs more precise

### PR DESCRIPTION
**Pull request by ohamel-softwaresecure**
Refactor: RAII-ify parser. Smaller, simpler.
Feature:
- Record nodes now have more precise location info.
- You can now `join(collection_of_ptrs)` and not get pointless nonsense. Pointer-like types automatically deref rather than print the pointer value (which is never what you want).

Other:
- some very early groundwork for future functionality (e.g. aggregates various arity)

War Crimes:
- Undead raii wrapper class `std::auto_ptr` returns from the grave, this time hiding behind its new name `souffle::parser::Mov`. Unfortunately the only way to banish this abomination is to move to Bison 3.2, where it finally gets its crap together and supports move semantics for semantic values. Prior to that semantic values *must* use copy semantics. (This is not a joke, it took ~8 years after C++11 for this to land in a release.)

### CODE REVIEW

File: src/Util.h [hunk 1]
```diff
@@ -472,8 +483,8 @@ std::unique_ptr<A> clone(const std::unique_ptr<A>& node) {
 }
 
 template <typename A>
-std::vector<std::unique_ptr<A>> clone(const std::vector<std::unique_ptr<A>>& xs) {
-    std::vector<std::unique_ptr<A>> ys;
+auto clone(const std::vector<A>& xs) {
```
Comment (mmcgr) [inline src/Util.h:486]: Can this be generalised to handle more than vectors? Vectors are easily our most common case so it's not an important change.

Comment (mmcgr) [inline src/Util.h:486]: While you're here, the variable names could be a lot more meaningful. `xs` ->`nodes` and `ys` -> `newNodes` maybe?

File: src/Util.h [hunk 2]
```diff
@@ -290,6 +290,17 @@ inline bool isNumber(const char* str) {
...
 // -------------------------------------------------------------------------------
 
+template <typename A>
+using Own = std::unique_ptr<A>;
+
+template <typename A>
+using VecOwn = std::vector<Own<A>>;
+
+template <typename A, typename B = A, typename... Args>
+Own<A> mk(Args&&... xs) {
```
Comment (mmcgr) [inline src/Util.h:300]: `mk` is not a helpful name. A more descriptive name would be good. As it is, I'm not sure what the function is for. A brief description would also help.

From the use in other files it seems to just be a way to shorten `std::unique_ptr` and `std::vector`, which makes it less clear what they are. Is there another advantage to these templates?

File: src/parser.yy [hunk 3]
```diff
@@ -67,6 +70,8 @@
             (Cur).filenames     = YYRHSLOC(Rhs, 0).filenames;   \
         }                                                       \
     } while (0)
+
+    using std::move;
```
Comment (mmcgr) [inline src/parser.yy:74]: I know std::move is now *everywhere*, but I think the extra 5 characters each time do make it clearer when scanning through the code. This does also get included in a header file, although currently that only contaminates main.cpp.

File: src/AstArgument.h [hunk 4]
```diff
@@ -41,11 +41,13 @@ namespace souffle {
  */
 class AstVariable : public AstArgument {
 public:
-    AstVariable(std::string name) : name(std::move(name)) {}
+    AstVariable(std::string name, SrcLocation loc = {}) : name(std::move(name)) {
+        setSrcLoc(std::move(loc));
+    }
```
Comment (mmcgr) [inline src/AstArgument.h:46]: I think these are an improvement, but would it perhaps be even better to push location up to the AstNode constructor?
```suggestion
    AstVariable(std::string name, SrcLocation loc = {}) : AstArgument(loc), name(std::move(name)) { }
```

File: src/Util.h (hunk 1 above)
Comment (ohamel-softwaresecure) [reply src/Util.h:486]: Yep, just add a new templated overload.

File: src/Util.h (hunk 2 above)
Comment (ohamel-softwaresecure) [reply src/Util.h:300]: 'mk' is shorthand for 'make'  (common idiom from Haskell leaking through), which given the ubiquity of the operation seemed to be appropriate to make quite short/easy to type/read. It's almost identical to `std::make_unique` except that you can optionally specify what supertype to use for the pointer parameterisation (useful when this interacts with type deduction, e.g. `auto`, lambdas, and template helpers).

File: src/parser.yy (hunk 3 above)
Comment (ohamel-softwaresecure) [reply src/parser.yy:74]: It makes some lines longer than 110 chars, forcing line wraps. I'll change it, but the inclusion in header was unintended.

Comment (ohamel-softwaresecure) [reply src/parser.yy:74]: Most of these are due to not having access to `api.cpp.automove`, which requires version 3.2. What platforms are holding us back? Have any updated recently?

File: src/Util.h (hunk 2 above)
Comment (ohamel-softwaresecure) [reply src/Util.h:300]: As for 'just to shorten it': It is. However, given how common this particular pattern is I find it more tiresome to read in its entirety (common things should have short names, much like words in natural languages, `std::vector<std::unique_ptr<A>>` vs `VecOwn<A>`). If this is a point of contention I'll revert.

**Comment by ohamel-softwaresecure** [comment]
Clang-10 apparently resolves templates slightly differently than gcc-9 and it's driving me up the wall.[/rant]

**Comment by ohamel-softwaresecure** [comment]
Cool beans. Bison 3.0.4 is a hot load of garbage that makes me very very mad. It is brain damaged and emits `value.copy< RuleBody > (that.value);` instead of `move<`. No possible workaround other than copying the `lalr1.cc` skeleton and patching it. This means that its impossible to safely* RAII-ify this junk with bison 3.0.4.

I quote https://fossies.org/linux/bison/ChangeLog :
```
10386 2018-09-13  Akim Demaille  <akim.demaille@gmail.com>
10387 
10388 	lalr1.cc: support move semantics
```
First release with this fix (according to changelog): `version 3.2.1`

Fine. Two can play stupid games. Hold my open sores for me.

File: src/parser.yy (hunk 3 above)
Comment (ohamel-softwaresecure) [reply src/parser.yy]: Apparently Ubuntu 18 is stuck on 3.0. That's at least one upper bound.

File: src/Util.h (hunk 1 above)
Comment (ohamel-softwaresecure) [reply src/Util.h:492]: Technically `A` doesn't need to be AST nodes. `souffle::clone` could have been more verbosely named `souffle::deepCopy`. (Right now there's some non-kosher things going on with the `std::unique_ptr<A>` instances where they blindly assume there's a `A::clone` to call, but that could/should be overloadable.)

File: src/parser.yy (hunk 3 above)
Comment (ohamel-softwaresecure) [reply src/parser.yy]: The `using std::move` is no more. It was replaced by the `Mov` abomination, which incidentally makes it almost ready as-is for `api.cpp.automove` whenever we move to Bison 3.2.

**Comment by mmcgr** [comment]
**Ubuntu**
20.04: 3.5.1
19.10: 3.4.1
18.04: 3.0.4
**Debian**
Buster: 3.3.2
Stretch: 3.0.4
**Fedora**
31: 3.4.1
30: 3.0.4
**CentOS**
7/8: 3.0.4
**FreeBSD**
11/12: 3.5.3
**OSX**
stock: 2.3
brew: 3.5.4

CentOS and bionic/stretch are the problem. The last two will be less important soon. CentOS will keep being a problem, but CentOS7 developers have needed to install the devtoolset to compile souffle anyway.

Did you consider using `shared_ptr`? I know I have, when trying to make the parser more RAII-ish, but I decided instead to wait until newer bison versions were more common. I underestimated the timescale.

**Comment by ohamel-softwaresecure** [comment]
I haven't, but I don't think it would work in this case unless we change the AST to use `shared_ptr`. Problem is we can't (AFAIK) convert a `shared_ptr` to a `unique_ptr` (there's no `shared_ptr::release`). It also wouldn't cover the case where we're `move`ing vectors around (they'd have to be wrapped).

**Pull request merged**